diffsmhm/loader.py

```diff
@@ -73,6 +79,366 @@ def wrap_to_local_volume_inplace(pos, cen, lbox):
+            # only keep columns we want
+            if key not in important_keys:
+                continue
+
+            # integer dtypes
+            if key in ("halo_id", "upid", "mmhid"):
+                dt = "i8"
+            else:
+                dt = "f4"
+            halos[key] = hdf[key][...][_host_mpeak_mask].astype(dt)
+
+    # if mmhid not known, find it
+    if "mmhid" not in halos.keys():
+
+        mmh_info = find_and_write_most_massive_hosts(halo_file, export=False)
+
+        halos["mmhid"] = mmh_info[0]
+
+        halos["mmh_x"] = mmh_info[1]
+        halos["mmh_y"] = mmh_info[2]
+        halos["mmh_z"] = mmh_info[3]
+
+    assert len((set(halos["halo_id"]))) == len(halos["halo_id"])
+
+    # assign each rank a chunk to then distribute and overload
+    # this is easier than only rank 0 loading for when we need to find mmh
+    avg, rem = divmod(len(halos["halo_id"]), N_RANKS)
+    rank_count = [avg + 1 if p < rem else avg for p in range(N_RANKS)]
+    displ = [sum(rank_count[:p]) for p in range(N_RANKS)]
+
+    rank_count = rank_count[RANK]
+    displ = displ[RANK]
+
+    halos_rank = {}
+    for key in halos.keys():
+        halos_rank[key] = halos[key][displ:displ+rank_count]
+
+    # munge
+    _munge_halos(halos_rank)
+
+    # fix "out of bounds" halos using periodicty
+    for pos in ["halo_x", "halo_y", "halo_z", "mmh_x", "mmh_y", "mmh_z"]:
+        halos_rank[pos][halos_rank[pos] < 0] += box_length
+        halos_rank[pos][halos_rank[pos] > box_length] -= box_length
+
+    # use MPIPartition to distribute and overload
+    partition = mpipartition.Partition()
+
+    halos_rank = mpipartition.distribute(partition, box_length, data=halos_rank,
+                                         coord_keys=("mmh_x", "mmh_y", "mmh_z"))
+    halos_rank["rank"] = np.zeros_like(halos_rank["halo_x"], dtype=np.int32) + RANK
+
+    # test: check that each partition has entire structure
+    mmhid_mod = np.copy(halos_rank["mmhid"])
+    mmhid_mod[mmhid_mod == -1] = halos_rank["halo_id"][halos_rank["mmhid"] == -1]
+
+    allhosts = np.unique(mmhid_mod)
+    neededsubs = halos["halo_id"][np.isin(halos["mmhid"], allhosts)]
+    assert len(np.setdiff1d(neededsubs, halos_rank["halo_id"])) == 0
+
+    halos_rank = mpipartition.overload(partition, box_length, halos_rank,
+                                       buff_wprp,
+                                       ["halo_x", "halo_y", "halo_z"],
+                                       structure_key="mmhid"
+                                       )
+
+    halos_rank["_inside_subvol"] = halos_rank["rank"] == RANK
+
+    # again, check that each rank has entire structure
+    mmhid_mod = np.copy(halos_rank["mmhid"])
+    mmhid_mod[mmhid_mod == -1] = halos_rank["halo_id"][halos_rank["mmhid"] == -1]
+
+    allhosts = np.unique(mmhid_mod)
+    neededsubs = halos["halo_id"][np.isin(halos["mmhid"], allhosts)]
+    assert len(np.setdiff1d(neededsubs, halos_rank["halo_id"])) == 0
```

Comment:
My current version of `find_and_write_most_massive_host()` also makes that assumption, which I realize is a problem. I can revisit finding a more distributed way, but without having spatially separated files I wasn't sure how to not load everything without scanning. 

Polaris is able to load the entire Bolshoi-Planck catalog fine with 1 rank per GPU. Since this is a loader method specifically for that data, do we need a different method currently?

I changed the way the catalog is initially loaded, so these checks are going to get moved to a test, but the "entire catalog on one process" problem is still relevant for loading everything with rank 0 and then distributing with mpipartition, and for finding each mmh.